XIPs/xip-35-message-sender-signature.md

```diff
@@ -0,0 +1,60 @@
+---
+xip: 35
+title: Message sender identifier in topic
+description: A system for detecting who sent a message in a topic from the Notification Server
+author: Noé Malzieu
+status: Draft
+type: Standards Track
+category: XRC
+created: 2023-12-06
+---
+
+
+## Abstract
+
+This XRC proposes to add a signature to public envelopes for messages, that a notification server could use to detect if one of the subscribed users has sent the message and not send him the notification (i.e. avoid receiving notifications for your own messages).
+
+## Motivation
+
+Right now, a message sent in a topic needs to have its payload decoded to understand who sent the message.
+When a notification server sees a message in a topic and decides to sent it to one of its subscribers, it has no way to know if the message is from the subscriber or not.
+This means a mobile XMTP client will subscribe to a topic then receive notifications for its own messages in the topic.
+This would be fine if the XMTP client could just drop the notification after having decoded the payload.
+On Apple devices, to be able to decode the payload before showing the notification, we must use a [Notification Service Extension](https://developer.apple.com/documentation/usernotifications/unnotificationserviceextension) - and the only way to drop a notification from a Notification Service Extension is to obtain a specific entitlement from Apple: [com.apple.developer.usernotifications.filtering](https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_usernotifications_filtering)
+Obtaining this entitlement from Apple has proven very hard - they might consider XMTP clients too web3, and they just don't provide this entitlement to apps that provide financial services.
+Specifically, Coinbase Wallet never managed to obtain this entitlement, and Converse had it for a few months before losing it.
+
+## Specification
+
+The SDKs would generate a private/public key pair for each topic.
+The private/public key pair should always be the same for a given user and a given topic - not dependent on installations / SDK language.
+
+Before publishing an envelope on a topic, the SDK would use the topic private key to sign the message payload and attach it to the envelope.
```

Comment:
It's important to note that the public key of the user's signing key would be private between the user's devices and the notification server. Third parties would not be able to validate that the signatures originate from the actual sender and were not forged.